lib/iris/fileformats/grib/_message.py

```diff
@@ -110,6 +108,8 @@ def data(self):
                 raise TranslationError(msg)
             if template == 90:
                 shape = (grid_section['Ny'], grid_section['Nx'])
+            elif template == 40 and reduced:
+                shape = (grid_section['numberOfDataPoints'],)
```

Comment:
A lat/lon grid (GDT0 and GDT1) can be reduced according to the GRIB2 spec.

I didn't like the way we do top-level checking for details that should be implemented by each template. For example, GDT0 can be reduced (valid in GRIB2) but if we don't support it (we don't) then it should be up to our implementation of GDT0 to say it doesn't support this, not part of the wider loading code.

This check is only here because the shape is different for reduced grids, it is not a validity check. I moved code that checks for unsupported reduced grids into the GDT0/1 template function.